### PR TITLE
redirect to static script based on gatekeeper/dcdo config [Delivers #109318408]

### PR DESCRIPTION
**Pull request by wjordan**
This PR will allow us to redirect the #reset action (e.g., https://studio.code.org/s/starwars/reset) to a DCDO-configured url, in order to dynamically offload Hour of Code traffic to a static site if needed.

### CODE REVIEW

File: dashboard/app/controllers/script_levels_controller.rb [hunk 1]
```diff
@@ -39,7 +39,9 @@ def reset
     # If the user is signed in, continue normally.
     redirect_path = build_script_level_path(@script.starting_level)
 
-    if current_user
+    if Gatekeeper.allows('redirect_static_script', where: {script_name: @script.name})
+      redirect_to(DCDO.get("static_script_#{@script.name}", redirect_path))
```
Comment (philbogle) [inline dashboard/app/controllers/script_levels_controller.rb:43]: This doesn't do the session state reset for for signed out users, does that matter?

If the whole site is melting down, might we want to redirect ALL users to a static site without explicitly having to set up a path for every one.

**Comment by philbogle** [comment]
Issue:  if the mysql database is unavailable, Rails unfortunately won't render pages even ones that don't rely on the database.  Also, if DCDO or Gatekeepr is unavailable then we won't be able to find the redirect path.

On consideration, I'm afraid we might need something lower level and more bulletproof that doesn't have Rails and database dependencies.

**Comment by wjordan** [comment]
The static site is only hosting the self-contained Hour of Code tutorials. If we did a full redirect (redirecting all users to static regardless of path), it would effectively take the rest of Code Studio outside of the Hour of Code tutorials offline. A partial redirect would leave the rest of Code Studio online while offloading the bulk of our Hour of Code traffic to the static site.

I would strongly prefer we have a more targeted path-based fallback available, versus the only option available being taking all of Code Studio offline with a full-site redirect. (The latter would be a small amount of work to have ready, so it's little extra effort to have both options ready.)

Assuming we still want a path-based fallback, here are our options for more work based on the concerns mentioned:
- If we want this to be resilient against Rails dependency failures, we can add the fallback redirection within a Rack middleware layer.
  - Rack middleware would still be able to use the dynamic config for on-the-fly adjustment.
  - If we don't trust the dynamic config system as a dependency (are there specific reasons why?), we don't have to use it to control switching the redirect on/off. However, we would then need to initiate a frontend-wide service update/restart of some sort to switch this.
- If we want this to be resilient against all Unicorn-based Ruby-app stack dependency failures, we can add the fallback redirection within Varnish. (This would preclude the ability to use the dynamic config for switching the redirect on/off.)
- Finally, if we want this to be completely resilient against our whole application stack, we could establish path-based redirections within a fallback CloudFront distribution (using path behaviors with custom origins pointing to the static site), that we switch to in the event of a failure.

**Comment by wjordan** [comment]
Closing as this feature is no longer needed post-HoC

**Pull request closed without merging**